guardian/pinboard
# Appsync integration

## What does this change?

This PR includes a number of changes:
1. introducing the GraphQL code generator, Apollo client, and using Apollo to add hooks to AWS AppSync GraphQL API
1. rendering the chat window with appropriately sorted messages and notification icons
1. adjusting the structure of the user data we store in our Dynamo table and configuring the table to use pay per request billing

Currently, we support only a single shared Pinboard, which can be accessed by running the service locally. Future work will introduce the necessary changes to support multiple Pinboards (one per Workflow item).

In more detail:

1. The [code generator](https://graphql-code-generator.com/) allows us to generate TypeScript types and (GraphQL) resolvers from our GraphQL schema. Apollo provides us with hooks to query, mutate, and subscribe to our AppSync API. Subscriptions allow us to receive real time updates whenever new messages are added to the Pinboard.

1. The widget can now be expanded into a functioning chat window, with messages sorted from oldest to newest. The window will conditionally scroll to the most recent message received, depending on the users position in the message history. The widget can also display two notification icons:  a dot to signify there are unread messages, and a warning symbol to signify there has been some kind of error. In the case of an error, the error text will be prominently displayed at the top of the Pinboard itself when expanded. This design is likely to change in the future.

1. Instead of storing all elements of the `User` model - including unused/unnecessary and potentially insecure values - we take only the `firstName`, `lastName`, and `email`. We also modify the CDK definition of the Items table, to use pay per request billing, instead of the default provisioned billing mode.

## How to test
Run the branch locally and visit https://pinboard.local.dev-gutools.co.uk/. Observe that you can read and send messages through the Pinboard widget. Try opening Pinboard in two browser tabs: one with the widget open, and one with it closed. Send a message in one tab and observe the notification icon appearing on the widget in the other.

## How can we measure success?

## Have we considered potential risks?

## Images
